python/ml4ir/base/model/calibrations/temperature_scaling.py

```diff
@@ -0,0 +1,247 @@
+def timed_execution(logger):
+    """ a context manager to compute the execution time of a process usually within
+     a 'with' statement
+    Parameter:
+        logger: logging.Logger object
+    """
+    t0 = time.time()
+    yield
+    dt = time.time() - t0
+    logger.info(f'Evaluation took: {dt} seconds')
+
+
+def np_value(tensor):
+    """Get numpy value out of possibly nested tuple of tensors.
+    Parameters:
+        tensor: TF.Tensor
+                a tensor object
+    """
+    if isinstance(tensor, tuple):
+        return type(tensor)(*(np_value(t) for t in tensor))
+    else:
+        return tensor.numpy()
+
+
+def run(optimizer, logger):
+    """Run an optimizer and measure it's evaluation time."""
+    optimizer()  # Warmup.
+    with timed_execution(logger):
+        result = optimizer()
+    return np_value(result)
+
+
+def accuracy(scores, labels):
+    """
+    returns accuracy
+    Parameters:
+        scores: tf.Tensor object that contains softmaxes
+        labels: tf.Tensor object that contains labels
+
+    """
+    acc = tf.keras.metrics.Accuracy()
+    acc.update_state(labels, tf.argmax(scores, axis=-1))
+    return acc.result().numpy()
+
+
+def dict_to_zipped_csv(data_dic, root_dir):
+    """
+    saves input dictionary to a zipped csv file
+    Parameters:
+        data_dic: input dict to be converted to  zipped csv file
+        path: path to save the output file
+    """
+    # creating zip dir
+    zip_dir_path = os.path.join(root_dir, TEMPERATURE_SCALE)
+    os.mkdir(zip_dir_path)
+    csv_path = os.path.join(zip_dir_path, f'{TEMPERATURE_SCALE}.csv')
+
+    # creating .csv & .zip files
+    pd.DataFrame.from_dict(data_dic).to_csv(csv_path, index=False)
+    shutil.make_archive(zip_dir_path, "zip", root_dir, TEMPERATURE_SCALE)
+
+    # removing the dir, keeping only the zip
+    shutil.rmtree(zip_dir_path)
+    return zip_dir_path
+
+
+def get_intermediate_model(model, scorer):
+    """
+    returns a tf.keras.models.Model copy of the relavence_model.model. This model must generate
+    logits (inputs of softmax).
+    Parameters:
+        model:
+        scorer:
+    """
+    # get  last layer's output  --> MUST **NOT** BE AN ACTIVATION (e.g. SOFTMAX) LAYER
+    final_layer_name = scorer.model_config['layers'][-1]['name']
+    layer_output = model.get_layer(name=final_layer_name).output
+
+    return tf.keras.models.Model(inputs=model.input, outputs=layer_output)
+
+
+def eval_relevance_model(scorer: ScorerBase, logits, labels, temperature=None):
+    """
+    evaluate the relevance model given the logits and labels
+    Parameters:
+        scorer: Scorer of the RelevanceModel
+        logits: numpy array, logits (input of softmax)
+        labels: numpy array, class labels
+        temperature: TF.Tensor, a parameter of size=(1)
+    Returns:
+        accuracy and NLL loss
+    """
+    # convert to TF.Tensor
+    logits_tensor = tf.constant(logits, name='logits_test', dtype=tf.float32)
```

Comment:
@marjanhs Why are we converting them to `tf.constant` here? Aren't they already tensors? This might make things slow. (Not 100% sure)

Reply:
If your goal is to cast you can do that with `tf.cast(..., name="...")`

Reply:
no they are not tensors. They (logits) are `numpy` arrays

Reply:
Definitely worth documenting the data type in the docstrings.

Reply:
yes, have added : https://github.com/salesforce/ml4ir/pull/115/files#diff-bf0198f6577050795b504acc0a05fccc56e8942b07df2a0497c4cc467bfefd39R172-R174